livekit-agents/livekit/agents/pipeline/pipeline_agent.py

```diff
@@ -1,1330 +1,664 @@
+    def create(allow_interruptions: bool = True, step_index: int = 0) -> SpeechHandle:
+        return SpeechHandle(
+            speech_id=utils.shortuuid("speech_"),
+            allow_interruptions=allow_interruptions,
+            step_index=step_index,
+        )
 
     @property
-    def agent(self) -> "VoicePipelineAgent":
-        return self._assistant
+    def id(self) -> str:
+        return self._id
 
     @property
-    def chat_ctx(self) -> ChatContext:
-        return self._llm_stream.chat_ctx
+    def step_index(self) -> int:
+        return self._step_index
 
-    def store_metadata(self, key: str, value: Any) -> None:
-        self._metadata[key] = value
+    @property
+    def interrupted(self) -> bool:
+        return self._interrupt_fut.done()
 
-    def get_metadata(self, key: str, default: Any = None) -> Any:
-        return self._metadata.get(key, default)
+    @property
+    def allow_interruptions(self) -> bool:
+        return self._allow_interruptions
 
-    def llm_stream(self) -> LLMStream:
-        return self._llm_stream
+    def play(self) -> None:
+        self._play_fut.set_result(None)
 
-    def add_extra_chat_message(self, message: ChatMessage) -> None:
-        """Append chat message to the end of function outputs for the answer LLM call"""
-        self._extra_chat_messages.append(message)
+    def done(self) -> bool:
+        return self._done_fut.done()
 
-    @property
-    def extra_chat_messages(self) -> list[ChatMessage]:
-        return self._extra_chat_messages
+    def interrupt(self) -> None:
+        if not self._allow_interruptions:
+            raise ValueError("This generation handle does not allow interruptions")
 
+        if self.done():
+            return
 
-def _default_before_llm_cb(
-    agent: VoicePipelineAgent, chat_ctx: ChatContext
-) -> LLMStream:
-    return agent.llm.chat(
-        chat_ctx=chat_ctx,
-        fnc_ctx=agent.fnc_ctx,
-    )
+        self._done_fut.set_result(None)
+        self._interrupt_fut.set_result(None)
 
+    async def wait_for_playout(self) -> None:
+        await asyncio.shield(self._playout_done_fut)
 
-@dataclass
-class SpeechData:
-    sequence_id: str
+    def _mark_playout_done(self) -> None:
+        self._playout_done_fut.set_result(None)
 
+    def _mark_done(self) -> None:
+        with contextlib.suppress(asyncio.InvalidStateError):
+            # will raise InvalidStateError if the future is already done (interrupted)
+            self._done_fut.set_result(None)
 
-SpeechDataContextVar = contextvars.ContextVar[SpeechData]("voice_assistant_speech_data")
 
-
-def _default_before_tts_cb(
-    agent: VoicePipelineAgent, text: str | AsyncIterable[str]
-) -> str | AsyncIterable[str]:
-    return text
+EventTypes = Literal[
+    "user_started_speaking",
+    "user_stopped_speaking",
+    "agent_started_speaking",
+    "agent_stopped_speaking",
+    "user_message_committed",
+    "agent_message_committed",
+    "agent_message_interrupted",
```

Comment:
we need to add metrics still?

what about these?
    "function_calls_collected",
    "function_calls_finished",